File: exercises/C_Basics/04_control_flow.c
/*
 * Exercises for Lesson 04: Control Flow
 * Topic: C_Basics
 * Solutions to practice problems from the lesson.
 * Compile: gcc -Wall -Wextra -std=c11 -o ex04 04_control_flow.c
 */
#include <stdio.h>

/* === Exercise 1: Pattern Printing === */
/* Problem: Print several patterns using nested loops. */
void exercise_1(void) {
    printf("=== Exercise 1: Pattern Printing ===\n");

    int n = 5;

    /* Right triangle */
    printf("Right triangle (n=%d):\n", n);
    for (int i = 1; i <= n; i++) {
        for (int j = 0; j < i; j++) {
            printf("* ");
        }
        printf("\n");
    }

    /* Inverted triangle */
    printf("\nInverted triangle:\n");
    for (int i = n; i >= 1; i--) {
        for (int j = 0; j < i; j++) {
            printf("* ");
        }
        printf("\n");
    }

    /* Diamond */
    printf("\nDiamond (n=%d):\n", n);
    for (int i = 1; i <= n; i++) {
        for (int j = 0; j < n - i; j++) printf(" ");
        for (int j = 0; j < 2 * i - 1; j++) printf("*");
        printf("\n");
    }
    for (int i = n - 1; i >= 1; i--) {
        for (int j = 0; j < n - i; j++) printf(" ");
        for (int j = 0; j < 2 * i - 1; j++) printf("*");
        printf("\n");
    }

    /* Number pyramid */
    printf("\nNumber pyramid:\n");
    for (int i = 1; i <= n; i++) {
        for (int j = 0; j < n - i; j++) printf(" ");
        for (int j = 1; j <= i; j++) printf("%d ", j);
        printf("\n");
    }
}

/* === Exercise 2: Menu System === */
/* Problem: Implement a menu-driven program using switch and do-while. */
void exercise_2(void) {
    printf("\n=== Exercise 2: Menu System ===\n");

    /*
     * In a real interactive program, this would use scanf in a loop.
     * Here we simulate menu choices to demonstrate the pattern.
     */
    int choices[] = {1, 2, 3, 4, 0};  /* simulated user input */
    int idx = 0;
    int choice;

    do {
        printf("\n--- Menu ---\n");
        printf("1. Celsius to Fahrenheit\n");
        printf("2. Fahrenheit to Celsius\n");
        printf("3. Km to Miles\n");
        printf("4. Miles to Km\n");
        printf("0. Quit\n");

        choice = choices[idx++];
        printf("Choice (simulated): %d\n", choice);

        switch (choice) {
            case 1: {
                double c = 100.0;
                printf("%.1f C = %.1f F\n", c, c * 9.0 / 5.0 + 32.0);
                break;
            }
            case 2: {
                double f = 212.0;
                printf("%.1f F = %.1f C\n", f, (f - 32.0) * 5.0 / 9.0);
                break;
            }
            case 3: {
                double km = 10.0;
                printf("%.1f km = %.2f miles\n", km, km * 0.621371);
                break;
            }
            case 4: {
                double mi = 6.0;
                printf("%.1f miles = %.2f km\n", mi, mi / 0.621371);
                break;
            }
            case 0:
                printf("Goodbye!\n");
                break;
            default:
                printf("Invalid choice.\n");
                break;
        }
    } while (choice != 0);
}

/* === Exercise 3: Number Classification === */
/* Problem: Classify numbers using various control flow constructs. */
void exercise_3(void) {
    printf("\n=== Exercise 3: Number Classification ===\n");

    int numbers[] = {0, 1, 2, 3, 7, 12, 15, 28, 37, 100};
    int count = sizeof(numbers) / sizeof(numbers[0]);

    for (int idx = 0; idx < count; idx++) {
        int n = numbers[idx];
        printf("\n%d: ", n);

        /* Even/odd */
        printf("%s", (n % 2 == 0) ? "even" : "odd");

        /* Prime check */
        if (n > 1) {
            int is_prime = 1;
            for (int i = 2; i * i <= n; i++) {
                if (n % i == 0) {
                    is_prime = 0;
                    break;
                }
            }
            if (is_prime) printf(", prime");
        }

        /* Perfect number check (sum of proper divisors equals number) */
        if (n > 1) {
            int div_sum = 0;
            for (int i = 1; i < n; i++) {
                if (n % i == 0) div_sum += i;
            }
            if (div_sum == n) printf(", perfect");
        }

        /* FizzBuzz classification */
        if (n > 0) {
            if (n % 15 == 0)      printf(", FizzBuzz");
            else if (n % 3 == 0)  printf(", Fizz");
            else if (n % 5 == 0)  printf(", Buzz");
        }
    }
    printf("\n");
}

int main(void) {
    exercise_1();
    exercise_2();
    exercise_3();

    printf("\nAll exercises completed!\n");
    return 0;
}

File: exercises/CSharp_Basics/04_control_flow.cs
/*
 * Exercises for Lesson 04: Control Flow
 * Topic: CSharp_Basics
 * Solutions to practice problems from the lesson.
 * Run: dotnet run
 */

// Exercise 1: Switch expressions with pattern matching
void Exercise1()
{
    Console.WriteLine("=== Exercise 1: Switch Expressions ===");

    // Classify HTTP status codes using switch expression
    int[] statusCodes = { 200, 201, 301, 400, 403, 404, 500, 503 };
    foreach (int code in statusCodes)
    {
        string description = code switch
        {
            200 => "OK",
            201 => "Created",
            301 => "Moved Permanently",
            400 => "Bad Request",
            403 => "Forbidden",
            404 => "Not Found",
            500 => "Internal Server Error",
            503 => "Service Unavailable",
            >= 200 and < 300 => "Success",
            >= 300 and < 400 => "Redirection",
            >= 400 and < 500 => "Client Error",
            >= 500 => "Server Error",
            _ => "Unknown"
        };
        Console.WriteLine($"  HTTP {code}: {description}");
    }
    Console.WriteLine();
}

// Exercise 2: Various loop patterns — for, foreach, while, do-while
void Exercise2()
{
    Console.WriteLine("=== Exercise 2: Loop Patterns ===");

    // FizzBuzz using a for loop
    Console.Write("FizzBuzz (1-20): ");
    for (int i = 1; i <= 20; i++)
    {
        string output = (i % 3, i % 5) switch
        {
            (0, 0) => "FizzBuzz",
            (0, _) => "Fizz",
            (_, 0) => "Buzz",
            _ => i.ToString()
        };
        Console.Write($"{output} ");
    }
    Console.WriteLine();

    // Collatz sequence using while
    int n = 27;
    int steps = 0;
    Console.Write($"Collatz({n}): {n}");
    while (n != 1)
    {
        n = n % 2 == 0 ? n / 2 : 3 * n + 1;
        steps++;
        if (steps <= 10) Console.Write($" -> {n}");
    }
    Console.WriteLine($" ... ({steps} steps total)");

    // Pyramid with nested loops
    Console.WriteLine("Pyramid:");
    int height = 5;
    for (int row = 1; row <= height; row++)
    {
        Console.Write(new string(' ', height - row));
        Console.WriteLine(new string('*', 2 * row - 1));
    }
    Console.WriteLine();
}

// Exercise 3: Pattern matching with is, when, and relational patterns
void Exercise3()
{
    Console.WriteLine("=== Exercise 3: Pattern Matching ===");

    object[] items = { 42, "hello", 3.14, null!, true, 'A', new int[] { 1, 2, 3 } };

    foreach (object? item in items)
    {
        string desc = item switch
        {
            int i when i > 0 => $"Positive int: {i}",
            int i => $"Non-positive int: {i}",
            string { Length: > 5 } s => $"Long string: \"{s}\"",
            string s => $"Short string: \"{s}\"",
            double d => $"Double: {d:F2}",
            bool b => $"Boolean: {b}",
            char c => $"Char: '{c}'",
            int[] arr => $"Int array with {arr.Length} elements",
            null => "null value",
            _ => $"Unknown type: {item.GetType().Name}"
        };
        Console.WriteLine($"  {desc}");
    }

    // Relational and logical patterns
    Console.WriteLine("\nBMI Classification:");
    double[] bmis = { 16.5, 18.5, 22.0, 27.0, 32.5 };
    foreach (double bmi in bmis)
    {
        string category = bmi switch
        {
            < 18.5 => "Underweight",
            >= 18.5 and < 25.0 => "Normal",
            >= 25.0 and < 30.0 => "Overweight",
            >= 30.0 => "Obese",
            _ => "Invalid"
        };
        Console.WriteLine($"  BMI {bmi:F1} -> {category}");
    }
    Console.WriteLine();
}

// Exercise 4: break, continue, and labeled loops
void Exercise4()
{
    Console.WriteLine("=== Exercise 4: Loop Control ===");

    // Skip even numbers, stop at 15
    Console.Write("Odd numbers (stop at 15): ");
    for (int i = 1; i <= 20; i++)
    {
        if (i > 15) break;
        if (i % 2 == 0) continue;
        Console.Write($"{i} ");
    }
    Console.WriteLine();

    // Find first prime pair (twin primes)
    Console.Write("Twin primes under 50: ");
    for (int i = 2; i < 50; i++)
    {
        if (IsPrime(i) && IsPrime(i + 2))
            Console.Write($"({i},{i + 2}) ");
    }
    Console.WriteLine();

    // Nested loop with early exit using a flag
    Console.WriteLine("Finding first pair (i,j) where i*j == 42:");
    bool found = false;
    for (int i = 1; i <= 10 && !found; i++)
    {
        for (int j = 1; j <= 10 && !found; j++)
        {
            if (i * j == 42)
            {
                Console.WriteLine($"  Found: {i} * {j} = 42");
                found = true;
            }
        }
    }
    Console.WriteLine();

    static bool IsPrime(int n)
    {
        if (n < 2) return false;
        for (int i = 2; i * i <= n; i++)
            if (n % i == 0) return false;
        return true;
    }
}

// Exercise 5: Implement a simple menu-driven calculator using do-while and switch
void Exercise5()
{
    Console.WriteLine("=== Exercise 5: Menu-Driven Calculator (Simulated) ===");

    // Simulate a sequence of operations
    (string op, double a, double b)[] operations =
    {
        ("+", 10, 3),
        ("-", 25, 7),
        ("*", 6, 8),
        ("/", 100, 3),
        ("/", 5, 0),
        ("%", 17, 5)
    };

    foreach (var (op, a, b) in operations)
    {
        string result = op switch
        {
            "+" => $"{a} + {b} = {a + b}",
            "-" => $"{a} - {b} = {a - b}",
            "*" => $"{a} * {b} = {a * b}",
            "/" when b != 0 => $"{a} / {b} = {a / b:F4}",
            "/" => $"{a} / {b} = Error (division by zero)",
            "%" when b != 0 => $"{a} % {b} = {a % b}",
            _ => $"Unknown operator: {op}"
        };
        Console.WriteLine($"  {result}");
    }
    Console.WriteLine();
}

// Main
Exercise1();
Exercise2();
Exercise3();
Exercise4();
Exercise5();
